pkg/gardenlet/controller/controllerinstallation/controllerinstallation/reconciler.go

```diff
@@ -326,25 +327,27 @@ func (r *Reconciler) delete(
 			Namespace: v1beta1constants.GardenNamespace,
 		},
 	}
-	if err := r.SeedClientSet.Client().Delete(seedCtx, mr); err == nil {
-		log.Info("Deletion of ManagedResource is still pending", "managedResource", client.ObjectKeyFromObject(mr))
 
-		msg := fmt.Sprintf("Deletion of ManagedResource %q is still pending.", controllerInstallation.Name)
-		conditionInstalled = v1beta1helper.UpdatedConditionWithClock(r.Clock, conditionInstalled, gardencorev1beta1.ConditionFalse, "DeletionPending", msg)
-		return reconcile.Result{RequeueAfter: RequeueDurationWhenResourceDeletionStillPresent}, nil
-	} else if !apierrors.IsNotFound(err) {
-		conditionInstalled = v1beta1helper.UpdatedConditionWithClock(r.Clock, conditionInstalled, gardencorev1beta1.ConditionFalse, "DeletionFailed", fmt.Sprintf("Deletion of ManagedResource %q failed: %+v", controllerInstallation.Name, err))
+	if err := client.IgnoreNotFound(managedresources.Delete(seedCtx, r.SeedClientSet.Client(), mr.Namespace, mr.Name, false)); err != nil {
+		log.Info("Deletion of ManagedResource and its secrets failed", "managedResource", client.ObjectKeyFromObject(mr))
+		conditionInstalled = v1beta1helper.UpdatedConditionWithClock(r.Clock, conditionInstalled, gardencorev1beta1.ConditionFalse, "DeletionFailed", fmt.Sprintf("Deletion of ManagedResource %q and its secrets failed: %+v", controllerInstallation.Name, err))
 		return reconcile.Result{}, err
 	}
 
-	secret := &corev1.Secret{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      controllerInstallation.Name,
-			Namespace: v1beta1constants.GardenNamespace,
-		},
-	}
-	if err := r.SeedClientSet.Client().Delete(seedCtx, secret); client.IgnoreNotFound(err) != nil {
-		conditionInstalled = v1beta1helper.UpdatedConditionWithClock(r.Clock, conditionInstalled, gardencorev1beta1.ConditionFalse, "DeletionFailed", fmt.Sprintf("Deletion of ManagedResource secret %q failed: %+v", controllerInstallation.Name, err))
+	fiveSecond := time.Second * 5
+	timeoutSeedCtx, cancelCtx := context.WithTimeout(seedCtx, fiveSecond)
+	defer cancelCtx()
+	if err := managedresources.WaitUntilDeleted(timeoutSeedCtx, r.SeedClientSet.Client(), mr.Namespace, mr.Name); err != nil {
```

Comment:
Why do we need to wait here and don't stick to the re-enqueuing approach?

Reply:
My first thought was to use the high level function but now that I think more about it a better approach might be

1. Try getting the mr
2. if there are no errors - requeue
3. if err is not found - continue with the flow
4. if err is something else - we got different problem at hand (log and return the error)

Is this what you are suggesting?

Reply:
That's also fine. What I rather meant is that the current implementation of the delete procedure should be good enough and we probably shouldn't add a waiting step just to extract errors from the ManagedResource. Operators can still look at the ManagedResource in order to find out what's wrong.

Reply:
Hm, but imo we still need to check if the mr is successfully deleted before continuing with the namespace deletion, right?

Reply:
Absolutely, and this already happens today (https://github.com/gardener/gardener/blob/ce96f9a15a81a2633f8d3ae64dccb1122460ec22/pkg/gardenlet/controller/controllerinstallation/controllerinstallation/reconciler.go#L330). So let me rephrase the quest: Why do we need to change something here in the scope of this PR? 😃

Reply:
Previously we were using `client.Delete()` which does not include all the cool logic behind `managedresources.Delete()`. So now when using `managedresources.Delete()` we still need to adapt the reconciler and ensure that the mr is actually gone before continuing.

I will make another commit with the suggested final proposal of this check